### PR TITLE
Formalize documentation format and content

### PR DESCRIPTION
**Pull request by pineapplemachine**
File `docs/documenting.md` is the most important addition - read it here:
https://github.com/pineapplemachine/higher/blob/08745f942aad3db078d6d9e5650788e324af8d62/docs/documenting.md

This formalizes the documentation format for which up to now I've just pretty much been winging it

It also happens to move `arrayAsSequence` and friends out of core/ and into functions/

### CODE REVIEW

**Comment by pineapplemachine** [comment]
I will also write formalizations like this for function wrapping, sequence and error and contract creation, when I feel that the formats are sufficiently complete and final.

This file and those files will be linked to in a contributing guide and all of them should eventually be available at the same site as the API documentation.

**Comment by pineapplemachine** [comment]
Before this is merged I think I'd like to try a slightly better pattern for core docs

**Comment by pineapplemachine** [comment]
Ok I think I'm done now code-wise

**Comment by jpsullivan** [comment]
Will definitely get to this asap. Lots to look over on this one.

File: src/core/sequence.js [hunk 1]
```diff
@@ -100,6 +153,7 @@ Sequence.prototype.root = function(){
 };
 
 // Get a string indicating the sequence type chain represented by sources.
+// TODO: This probably doesn't need to be in the prototype, honestly
```
Comment (jpsullivan) [inline src/core/sequence.js:156]: Probably doesn't, however I could see some potential use-cases for debugging if for whatever reason the type chain needs to be verified.

**Review by jpsullivan — APPROVED** [review]
Sorry about the delay getting this reviewed. Been crazy busy lately.

Overall this is looking good. I like the introduction of `lightWrap` as a lightweight alternative for internal components. However, I see no mention of what kind of performance issues you were seeing with `wrap`. Either way, I trust you didn't create `lightWrap` for no reason, just interested s'all.

Otherwise, all looks good here. Thanks for the patience 👍

File: src/core/sequence.js (hunk 1 above)
Comment (pineapplemachine) [reply src/core/sequence.js:156]: At some point I want to move it to a function that just gets a type string from whatever value. The only place I use this is for error messages and most of them have to account for other types also.

**Pull request merged**